### ISSUE TITLE
fake_trampolines test ERROR

### ISSUE THREAD

**Issue by Mdyrl7777777** [smeso/sara-test#2]
```
$ sara-test
These tests should pass even with SARA disabled:
              wx_mappings:      OK
             nx_shellcode:      OK
     fake_trampoline_heap:      OK

These tests should pass with SARA fully enabled:
             anon_mmap_wx:      OK
             file_mmap_wx:      OK
     gnu_executable_stack:      OK
            heap_mprotect:      OK
           stack_mprotect:      OK
       anon_mmap_mprotect:      OK
       file_mmap_mprotect:      OK
            text_mprotect:      OK
             bss_mprotect:      OK
            data_mprotect:      OK
                mmap_exec:      OK
                 transfer:      OK
 gcc_trampolines_working1:      OK
 gcc_trampolines_working2:      OK
         fake_trampolines:      ERROR

Tests for procattr interface:
         correct_settings:      OK
         verbosity_change:      OK
          complain_change:      OK
     full_change_no_force:      OK
              force_wxorx:      OK
```
```
kernel: SARA: WXP: '/usr/bin/sara-test' run with flags '0x2f'.
kernel: sara-test[25111]: segfault at 7f78bea05e00 ip 00007f78bea05e00 sp 00007f78bea05de8 error 15
systemd[1]: Started Process Core Dump (PID 25112/UID 0).
kernel: SARA: WXP: '/usr/lib/sara-test/fake_tramp' run with flags '0x12f'.
kernel: fake_tramp[25116]: segfault at d36ce964b20 ip 00000d36ce964b20 sp 00007d97fca2e698 error 15
systemd[1]: Started Process Core Dump (PID 25117/UID 0).
systemd-coredump[25113]: Process 25111 (sara-test) of user 1000 dumped core.
                                                
                                                Stack trace of thread 25111:
                                                #0  0x00007f78bea05e00 n/a (n/a)
kernel: pr_wxp: 2 callbacks suppressed
kernel: SARA: WXP: W^X in '/usr/bin/sara-test' (25124).
kernel: SARA: WXP: W^X in '/usr/bin/sara-test' (25125).
kernel: SARA: WXP: '/usr/lib/sara-test/trampoline_exstack' run with flags '0x12f'.
kernel: SARA: WXP: W^X in '/usr/lib/sara-test/trampoline_exstack' (25127).
kernel: SARA: WXP: mprotect on anon mmap in '/usr/bin/sara-test' (25128).
kernel: SARA: WXP: mprotect on stack in '/usr/bin/sara-test' (25129).
kernel: SARA: WXP: mprotect on anon mmap in '/usr/bin/sara-test' (25130).
kernel: SARA: WXP: mprotect on file mmap in '/usr/bin/sara-test' (25131).
kernel: SARA: WXP: W^X in '/usr/bin/sara-test' (25132).
kernel: SARA: WXP: mprotect on anon mmap in '/usr/bin/sara-test' (25133).
kernel: SARA: WXP: mprotect on file mmap in '/usr/bin/sara-test' (25134).
kernel: SARA: WXP: '/usr/lib/sara-test/transfer' run with flags '0x2f'.
kernel: SARA: WXP: '/usr/lib/sara-test//transfer' run with flags '0x22f'.
kernel: SARA: WXP: '/usr/lib/sara-test/trampoline' run with flags '0x12f'.
kernel: SARA: WXP: '/usr/lib/sara-test/trampoline_nopie' run with flags '0x12f'.
kernel: SARA: WXP: '/usr/lib/sara-test/fake_tramp' run with flags '0x12f'.
kernel: fake_tramp[25142]: segfault at 7957e800 ip 000000007957e800 sp 00007f4f8ca19ce8 error 14 in fake_tramp[f017957e000+1000]
systemd[1]: Started Process Core Dump (PID 25143/UID 0).
systemd-coredump[25118]: Process 25116 (fake_tramp) of user 1000 dumped core.
                                                
                                                Stack trace of thread 25116:
                                                #0  0x00000d36ce964b20 n/a (n/a)
kernel: SARA: WXP: '/usr/lib/sara-test/procattr' run with flags '0x7c'.
systemd-coredump[25144]: Process 25142 (fake_tramp) of user 1000 dumped core.
                                                
                                                Stack trace of thread 25142:
                                                #0  0x000000007957e800 n/a (n/a)
```

**Comment by smeso** [comment]
Can you please disable systemd's coredump feature and try again?
Thank you for you help!

**Comment by Mdyrl7777777** [comment]
It looks similar:
```
kernel: SARA: WXP: '/usr/bin/sara-test' run with flags '0x2f'.
kernel: sara-test[4061]: segfault at 74e400ef7040 ip 000074e400ef7040 sp 000074e400ef7028 error 15
kernel: SARA: WXP: '/usr/lib/sara-test/fake_tramp' run with flags '0x12f'.
kernel: fake_tramp[4064]: segfault at 5481f2d4a90 ip 000005481f2d4a90 sp 000071aa534a1c48 error 15
kernel: pr_wxp: 2 callbacks suppressed
kernel: SARA: WXP: W^X in '/usr/bin/sara-test' (4066).
kernel: SARA: WXP: W^X in '/usr/bin/sara-test' (4067).
kernel: SARA: WXP: '/usr/lib/sara-test/trampoline_exstack' run with flags '0x12f'.
kernel: SARA: WXP: W^X in '/usr/lib/sara-test/trampoline_exstack' (4069).
kernel: SARA: WXP: mprotect on anon mmap in '/usr/bin/sara-test' (4070).
kernel: SARA: WXP: mprotect on stack in '/usr/bin/sara-test' (4071).
kernel: SARA: WXP: mprotect on anon mmap in '/usr/bin/sara-test' (4072).
kernel: SARA: WXP: mprotect on file mmap in '/usr/bin/sara-test' (4073).
kernel: SARA: WXP: W^X in '/usr/bin/sara-test' (4074).
kernel: SARA: WXP: mprotect on anon mmap in '/usr/bin/sara-test' (4075).
kernel: SARA: WXP: mprotect on file mmap in '/usr/bin/sara-test' (4076).
kernel: SARA: WXP: '/usr/lib/sara-test/transfer' run with flags '0x2f'.
kernel: SARA: WXP: '/usr/lib/sara-test//transfer' run with flags '0x22f'.
kernel: SARA: WXP: '/usr/lib/sara-test/trampoline' run with flags '0x12f'.
kernel: SARA: WXP: '/usr/lib/sara-test/trampoline_nopie' run with flags '0x12f'.
kernel: SARA: WXP: '/usr/lib/sara-test/fake_tramp' run with flags '0x12f'.
kernel: fake_tramp[4084]: segfault at 70775800 ip 0000000070775800 sp 000079e6a94c9ec8 error 14 in fake_tramp[2f970775000+1000]
kernel: SARA: WXP: '/usr/lib/sara-test/procattr' run with flags '0x7c'.
```

**Comment by Mdyrl7777777** [comment]
Backtrace `/usr/lib/sara-test/fake_tramp` from dbg:
```
#0  0x0000000000000800 in ?? ()
#1  0x000000010000067c in main (argc=<optimized out>, argv=<optimized out>) at src/fake_tramp.c:222
```

**Comment by smeso** [comment]
I suppose that it crashes even with SARA completely disabled, doesn't it?
If so, what compiler are you using? Can you send be the fake_tramp binary compiled with "-g" ?

**Comment by Mdyrl7777777** [comment]
With sara disabled there are more errors:
```
These tests should pass even with SARA disabled:
              wx_mappings:      OK
             nx_shellcode:      OK
     fake_trampoline_heap:      OK

These tests should pass with SARA fully enabled:
             anon_mmap_wx:      VULNERABLE
             file_mmap_wx:      VULNERABLE
     gnu_executable_stack:      VULNERABLE
            heap_mprotect:      ERROR
           stack_mprotect:      VULNERABLE
       anon_mmap_mprotect:      VULNERABLE
       file_mmap_mprotect:      VULNERABLE
            text_mprotect:      VULNERABLE
             bss_mprotect:      VULNERABLE
            data_mprotect:      VULNERABLE
                mmap_exec:      ERROR
                 transfer:      VULNERABLE
 gcc_trampolines_working1:      ERROR
 gcc_trampolines_working2:      ERROR
         fake_trampolines:      ERROR

Tests for procattr interface:
         correct_settings:      VULNERABLE
procattr tests disabled
```
```
kernel: sara-test[9612]: segfault at 74c820fd9920 ip 000074c820fd9920 sp 000074c820fd9908 error 15
kernel: fake_tramp[9615]: segfault at 7a7b0244c80 ip 000007a7b0244c80 sp 0000707a147ae008 error 15
kernel: sara-test[9621]: segfault at 76202638538 ip 00006d507f574661 sp 000074c820fd8788 error 7 in libc-2.26.so[6d507f4f6000+1a
kernel: trampoline[9633]: segfault at 7cef60c16604 ip 00007cef60c16604 sp 00007cef60c165f8 error 15
kernel: trampoline_nopi[9636]: segfault at 76d448d63dd4 ip 000076d448d63dd4 sp 000076d448d63dc8 error 15
kernel: fake_tramp[9638]: segfault at 7f8954d0d050 ip 00007f8954d0d050 sp 00007f8954d0d008 error 15

```

```
$ gcc -v
Target: x86_64-pc-linux-gnu
Configured with: /build/gcc/src/gcc/configure --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=https://bugs.archlinux.org/ --enable-languages=c,c++,ada,fortran,go,lto,objc,obj-c++ --enable-shared --enable-threads=posix --enable-libmpx --with-system-zlib --with-isl --enable-__cxa_atexit --disable-libunwind-exceptions --enable-clocale=gnu --disable-libstdcxx-pch --disable-libssp --enable-gnu-unique-object --enable-linker-build-id --enable-lto --enable-plugin --enable-install-libiberty --with-linker-hash-style=gnu --enable-gnu-indirect-function --disable-multilib --disable-werror --enable-checking=release --enable-default-pie --enable-default-ssp
Thread model: posix
gcc version 7.2.0 (GCC) 
```
Warning while building with gcc:
```
gcc -c -o src/sara-test.o src/sara-test.c -O2 -fPIE -fstack-protector -ggdb -DEXTRA_BINS_PATH=\"/usr/lib/sara-test/\"
gcc -c -o src/utils.o src/utils.c -O2 -fPIE -fstack-protector -ggdb -DEXTRA_BINS_PATH=\"/usr/lib/sara-test/\"
export CROSS_COMPILE
gcc -c -o src/trampoline.o src/trampoline.c -O2 -fPIE -fstack-protector -ggdb -DEXTRA_BINS_PATH=\"/usr/lib/sara-test/\"
cd ./src//libsara && make libsara.o
make[1]: Entering directory '/var/local/makepkg/build/build/sara-test/src/sara-test/src/libsara'
gcc -c -o libsara.o libsara.c -O2 -fstack-protector -fPIC -O2 -fPIE -fstack-protector -ggdb
libsara.c: In function ‘__get_wxprot_flags’:
libsara.c:40:20: warning: implicit declaration of function ‘strtoul’; did you mean ‘stdout’? [-Wimplicit-function-declaration]
  return (uint16_t) strtoul(buf, NULL, 16);
                    ^~~~~~~
                    stdout
src/utils.c: In function ‘get_perms’:
src/utils.c:44:24: warning: ‘sizeof’ on array function parameter ‘perms’ will return size of ‘char *’ [-Wsizeof-array-argument]
  snprintf(perms, sizeof(perms), "none");
                        ^
src/utils.c:29:47: note: declared here
 void get_perms(const void *m, pid_t pid, char perms[5])
                                               ^~~~~
src/utils.c:49:26: warning: ‘sizeof’ on array function parameter ‘perms’ will return size of ‘char *’ [-Wsizeof-array-argument]
    snprintf(perms, sizeof(perms), "%s", token);
                          ^
src/utils.c:29:47: note: declared here
 void get_perms(const void *m, pid_t pid, char perms[5])
                                               ^~~~~
src/utils.c:54:35: warning: ‘sizeof’ on array function parameter ‘perms’ will return size of ‘char *’ [-Wsizeof-array-argument]
  if (strncmp(perms, "none", sizeof(perms)) == 0) {
                                   ^
src/utils.c:29:47: note: declared here
 void get_perms(const void *m, pid_t pid, char perms[5])
                                               ^~~~~
gcc -c -o src/trampoline_nopie.o src/trampoline.c -O2 -O2 -fPIE -fstack-protector -ggdb -DEXTRA_BINS_PATH=\"/usr/lib/sara-test/\"
make[1]: Leaving directory '/var/local/makepkg/build/build/sara-test/src/sara-test/src/libsara'
gcc -z execstack -o bin/trampoline_exstack src/trampoline.o -Wl,-z,relro -Wl,-z,now -Wl,-Bsymbolic-functions -pie 
gcc -c -o src/procattr.o src/procattr.c -O2 -fPIE -fstack-protector -ggdb -DEXTRA_BINS_PATH=\"/usr/lib/sara-test/\"
gcc -c -o src/fake_tramp.o src/fake_tramp.c -Wno-pointer-to-int-cast -O2 -O2 -fPIE -fstack-protector -ggdb
gcc -c -o src/transfer.o src/transfer.c -O2 -fPIE -fstack-protector -ggdb -DEXTRA_BINS_PATH=\"/usr/lib/sara-test/\"
gcc -z noexecstack -o bin/trampoline src/trampoline.o -Wl,-z,relro -Wl,-z,now -Wl,-Bsymbolic-functions -pie 
gcc -z noexecstack -o bin/trampoline_nopie src/trampoline_nopie.o -fno-pie -O2 -fPIE -fstack-protector -ggdb
gcc -z noexecstack -o bin/fake_tramp src/fake_tramp.o -fno-pie -O2 -fPIE -fstack-protector -ggdb
gcc -o bin/transfer src/transfer.o src/libsara/libsara.o -Wl,-z,relro -Wl,-z,now -Wl,-Bsymbolic-functions -pie 
gcc -o bin/procattr src/procattr.o src/utils.o src/libsara/libsara.o -Wl,-z,relro -Wl,-z,now -Wl,-Bsymbolic-functions -pie 
gcc -o bin/sara-test src/sara-test.o src/utils.o src/libsara/libsara.o -Wl,-z,relro -Wl,-z,now -Wl,-Bsymbolic-functions -pie 
```

```
$ clang -v
clang version 5.0.0 (tags/RELEASE_500/final)
Target: x86_64-unknown-linux-gnu
Thread model: posix
```
Building with clang fails:
```
clang -c -o src/sara-test.o src/sara-test.c -O2 -fPIE -fstack-protector -ggdb -DEXTRA_BINS_PATH=\"/usr/lib/sara-test/\"
clang -c -o src/utils.o src/utils.c -O2 -fPIE -fstack-protector -ggdb -DEXTRA_BINS_PATH=\"/usr/lib/sara-test/\"
export CROSS_COMPILE
clang -c -o src/trampoline.o src/trampoline.c -O2 -fPIE -fstack-protector -ggdb -DEXTRA_BINS_PATH=\"/usr/lib/sara-test/\"
cd ./src//libsara && make libsara.o
make[1]: Entering directory '/var/local/makepkg/build/build/sara-test/src/sara-test/src/libsara'
clang -c -o libsara.o libsara.c -O2 -fstack-protector -fPIC -O2 -fPIE -fstack-protector -ggdb
src/trampoline.c:34:23: error: function definition is not allowed here
        void f2(void) {
                      ^
src/trampoline.c:37:5: error: use of undeclared identifier 'f2'
        f0(f2);
           ^
2 errors generated.
make: *** [Makefile:48: src/trampoline.o] Error 1
make: *** Waiting for unfinished jobs....
src/utils.c:37:39: warning: format specifies type 'unsigned short' but the argument has type 'pid_t' (aka 'int') [-Wformat]
        snprintf(path, 25, "/proc/%hu/maps", pid);
                                  ~~~        ^~~
                                  %d                                                                                                                   
src/utils.c:44:24: warning: sizeof on array function parameter will return size of 'char *' instead of 'char [5]' [-Wsizeof-array-argument]
        snprintf(perms, sizeof(perms), "none");
                              ^
src/utils.c:29:47: note: declared here
void get_perms(const void *m, pid_t pid, char perms[5])
                                              ^
src/utils.c:49:26: warning: sizeof on array function parameter will return size of 'char *' instead of 'char [5]' [-Wsizeof-array-argument]
                        snprintf(perms, sizeof(perms), "%s", token);
                                              ^
src/utils.c:29:47: note: declared here
void get_perms(const void *m, pid_t pid, char perms[5])
                                              ^
src/utils.c:54:35: warning: sizeof on array function parameter will return size of 'char *' instead of 'char [5]' [-Wsizeof-array-argument]
        if (strncmp(perms, "none", sizeof(perms)) == 0) {
                                         ^
src/utils.c:29:47: note: declared here
void get_perms(const void *m, pid_t pid, char perms[5])
                                              ^
src/utils.c:54:36: warning: 'strncmp' call operates on objects of type 'char' while the size is based on a different type 'char *'
      [-Wsizeof-pointer-memaccess]
        if (strncmp(perms, "none", sizeof(perms)) == 0) {
                    ~~~~~                 ^~~~~
src/utils.c:54:36: note: did you mean to provide an explicit length?
        if (strncmp(perms, "none", sizeof(perms)) == 0) {
                                          ^~~~~
src/utils.c:68:39: warning: format specifies type 'unsigned short' but the argument has type 'pid_t' (aka 'int') [-Wformat]
        snprintf(path, 25, "/proc/%hu/maps", pid);
                                  ~~~        ^~~
                                  %d                                                                                                                   
libsara.c:40:20: warning: implicit declaration of function 'strtoul' is invalid in C99 [-Wimplicit-function-declaration]
        return (uint16_t) strtoul(buf, NULL, 16);
                          ^
src/sara-test.c:48:47: warning: unknown attribute 'optimize' ignored [-Wunknown-attributes]
void __attribute__((noinline)) __attribute__((optimize("O0"))) __helper(void)
                                              ^
6 warnings generated.
1 warning generated.

```

**Comment by smeso** [comment]
I think this is fixed by a41e61cf873f316b33c14eaccce0b6cc79e9f75d can you confirm?
Also, please note that clang is not supported.

**Comment by Mdyrl7777777** [comment]
Thanks for the fixes. I used clang to show alternative source of information. There are no warnings during build with gcc now. (clang still doesn't build).

**sara-test (sara enabled):**
```
These tests should pass even with SARA disabled:
              wx_mappings:      OK
             nx_shellcode:      OK
     fake_trampoline_heap:      OK

These tests should pass with SARA fully enabled:
             anon_mmap_wx:      OK
             file_mmap_wx:      OK
     gnu_executable_stack:      OK
            heap_mprotect:      OK
           stack_mprotect:      OK
       anon_mmap_mprotect:      OK
       file_mmap_mprotect:      OK
            text_mprotect:      OK
             bss_mprotect:      OK
            data_mprotect:      OK
                mmap_exec:      OK
                 transfer:      OK
 gcc_trampolines_working1:      OK
 gcc_trampolines_working2:      OK
         fake_trampolines:      OK

Tests for procattr interface:
         correct_settings:      OK
         verbosity_change:      OK
          complain_change:      OK
     full_change_no_force:      OK
              force_wxorx:      OK
```
**journal log:**
```
kernel: SARA: WXP: '/usr/bin/sara-test' run with flags '0x2f'.
kernel: sara-test[9689]: segfault at 723b7715b370 ip 0000723b7715b370 sp 0000723b7715b358 error 15
kernel: SARA: WXP: '/usr/lib/sara-test/fake_tramp' run with flags '0x12f'.
kernel: fake_tramp[9692]: segfault at 3e6a6dd0 ip 000000003e6a6dd0 sp 000074da21db0358 error 15
kernel: pr_wxp: 2 callbacks suppressed
kernel: SARA: WXP: W^X in '/usr/bin/sara-test' (9694).
kernel: SARA: WXP: W^X in '/usr/bin/sara-test' (9695).
kernel: SARA: WXP: '/usr/lib/sara-test/trampoline_exstack' run with flags '0x12f'.
kernel: SARA: WXP: W^X in '/usr/lib/sara-test/trampoline_exstack' (9697).
kernel: SARA: WXP: mprotect on anon mmap in '/usr/bin/sara-test' (9698).
kernel: SARA: WXP: mprotect on stack in '/usr/bin/sara-test' (9699).
kernel: SARA: WXP: mprotect on anon mmap in '/usr/bin/sara-test' (9700).
kernel: SARA: WXP: mprotect on file mmap in '/usr/bin/sara-test' (9701).
kernel: SARA: WXP: W^X in '/usr/bin/sara-test' (9702).
kernel: SARA: WXP: mprotect on anon mmap in '/usr/bin/sara-test' (9703).
kernel: SARA: WXP: mprotect on file mmap in '/usr/bin/sara-test' (9704).
kernel: SARA: WXP: '/usr/lib/sara-test/transfer' run with flags '0x2f'.
kernel: SARA: WXP: '/usr/lib/sara-test//transfer' run with flags '0x22f'.
kernel: SARA: WXP: '/usr/lib/sara-test/trampoline' run with flags '0x12f'.
kernel: SARA: WXP: '/usr/lib/sara-test/trampoline_nopie' run with flags '0x12f'.
kernel: SARA: WXP: '/usr/lib/sara-test/fake_tramp' run with flags '0x12f'.
kernel: SARA: WXP: '/usr/lib/sara-test/procattr' run with flags '0x7c'.
```
**sara-test (sara disabled):**
```
These tests should pass even with SARA disabled:
              wx_mappings:      OK
             nx_shellcode:      OK
     fake_trampoline_heap:      OK

These tests should pass with SARA fully enabled:
             anon_mmap_wx:      VULNERABLE
             file_mmap_wx:      VULNERABLE
     gnu_executable_stack:      VULNERABLE
            heap_mprotect:      ERROR # it didn't show an error in one test but can't reproduce this
           stack_mprotect:      VULNERABLE
       anon_mmap_mprotect:      VULNERABLE
       file_mmap_mprotect:      VULNERABLE
            text_mprotect:      VULNERABLE
             bss_mprotect:      VULNERABLE
            data_mprotect:      VULNERABLE
                mmap_exec:      VULNERABLE
                 transfer:      VULNERABLE
 gcc_trampolines_working1:      ERROR
 gcc_trampolines_working2:      ERROR
         fake_trampolines:      ERROR

Tests for procattr interface:
         correct_settings:      VULNERABLE
procattr tests disabled
```
**journal log:**
```
kernel: sara-test[10529]: segfault at 76ba5eefb2f0 ip 000076ba5eefb2f0 sp 000076ba5eefb2d8 error 15
kernel: fake_tramp[10532]: segfault at ce4ee20 ip 000000000ce4ee20 sp 000070d0a74279f8 error 15
kernel: sara-test[10538]: segfault at d496c87e1f8 ip 0000604f613f5661 sp 000076ba5eefa148 error 7 in libc-2.26.so[604f61377000+1
kernel: trampoline[10550]: segfault at 74a35f24d2d4 ip 000074a35f24d2d4 sp 000074a35f24d2c8 error 15
kernel: trampoline_nopi[10553]: segfault at 7f151a083294 ip 00007f151a083294 sp 00007f151a083288 error 15
kernel: fake_tramp[10555]: segfault at 7a68fa05ec00 ip 00007a68fa05ec00 sp 00007a68fa05ebb8 error 15
kernel: sara-test[10676]: segfault at 7759470f7870 ip 00007759470f7870 sp 00007759470f7858 error 15
kernel: fake_tramp[10679]: segfault at 2dd98620 ip 000000002dd98620 sp 00007cba3ae73248 error 15
kernel: sara-test[10685]: segfault at ae30956d308 ip 0000650ee7b23661 sp 00007759470f66c8 error 7 in libc-2.26.so[650ee7aa5000+1
kernel: trampoline[10697]: segfault at 71f75e6a0344 ip 000071f75e6a0344 sp 000071f75e6a0338 error 15
kernel: trampoline_nopi[10700]: segfault at 7c4e8bb9baa4 ip 00007c4e8bb9baa4 sp 00007c4e8bb9ba98 error 15
kernel: fake_tramp[10702]: segfault at 73437e74d560 ip 000073437e74d560 sp 000073437e74d518 error 15
```
If above are expected please close this issue.

**Comment by smeso** [comment]
The error of heap_mprotect should be fixed by e0f7636331e701ce2d7d9a9f9f310112c2325ba5, anything else is fine. When sara is not enabled, trampolines' test are supposed to fail. They say "error" instead of "vulnerable" because it isn't a security issue.

**Comment by Mdyrl7777777** [comment]
All clear now. Thanks for help.
